crates/sui-graphql-rpc/src/types/transaction_block/mod.rs

```diff
@@ -297,102 +263,114 @@ impl TransactionBlock {
     ///
     /// If the `Page<Cursor>` is set, then this function will defer to the `checkpoint_viewed_at` in
     /// the cursor if they are consistent.
+    ///
+    /// Filters that involve a combination of `recvAddress`, `inputObject`, `changedObject`, and
+    /// `function` should provide a value for `scan_limit`. This indicates how many transactions to
+    /// scan through per the filter conditions.
     pub(crate) async fn paginate(
-        db: &Db,
+        ctx: &Context<'_>,
         page: Page<Cursor>,
         filter: TransactionBlockFilter,
         checkpoint_viewed_at: u64,
+        scan_limit: Option<u64>,
     ) -> Result<Connection<String, TransactionBlock>, Error> {
-        use transactions as tx;
+        if filter.is_empty() {
+            return Ok(Connection::new(false, false));
+        }
+
+        // If there is more than one `complex_filter` specified, then the caller has provided some
+        // arbitrary combination of `function`, `kind`, `recvAddress`, `inputObject`, or
+        // `changedObject`. Consequently, we require setting a `scanLimit`, or else we will return
+        // an error.
+        if filter.complex_filters() >= 2 && scan_limit.is_none() {
+            return Err(Error::Client(
+                "A scan limit must be specified for filter combinations involving `function`, `kind`, `recvAddress`, `inputObject`, or `changedObject`".to_string(),
+            ));
+        }
+
+        let limits = ctx.data_unchecked::<ServiceConfig>().limits;
+        if let Some(tx_ids) = filter.transaction_ids.as_ref() {
+            if tx_ids.len() > limits.max_transaction_ids as usize {
+                return Err(Error::Client(format!(
+                    "Transaction ids exceed max limit of {}",
```

Comment:
```suggestion
                    "Transaction IDs exceed max limit of {}",
```

Also, quote the actual number provided.